src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetHttpCacheUsageCollector.cs

```diff
@@ -0,0 +1,101 @@
+
+        public NuGetHttpCacheUsageCollector()
+        {
+            _vsInstanceData = new InstanceData();
+            ProtocolDiagnostics.HttpCacheHitEvent += HttpCacheUsage_CacheHitEventHandler;
+            ProtocolDiagnostics.HttpCacheMissEvent += HttpCacheUsage_CacheMissEventHandler;
+            InstanceCloseTelemetryEmitter.AddEventsOnShutdown += HttpCacheUsage_VSInstanceCloseHandler;
+        }
+
+        private void HttpCacheUsage_CacheHitEventHandler()
+        {
+            _vsInstanceData.IncrementCacheHitCount();
+        }
+
+        private void HttpCacheUsage_CacheMissEventHandler(bool cacheByPass, bool isExpired)
+        {
+            _vsInstanceData.IncrementCacheMissCount(cacheByPass, isExpired);
+        }
+
+        private void HttpCacheUsage_VSInstanceCloseHandler(object sender, TelemetryEvent telemetryEvent)
+        {
+            lock (_lock)
+            {
+                // Add VS Instance telemetry
+                _vsInstanceData.AddProperties(telemetryEvent);
+            }
+        }
+
+        public void Dispose()
+        {
+            ProtocolDiagnostics.HttpCacheHitEvent -= HttpCacheUsage_CacheHitEventHandler;
+            ProtocolDiagnostics.HttpCacheMissEvent -= HttpCacheUsage_CacheMissEventHandler;
+            InstanceCloseTelemetryEmitter.AddEventsOnShutdown -= HttpCacheUsage_VSInstanceCloseHandler;
+        }
+
+        internal class InstanceData
+        {
+            private int _hitCount = 0;
+            internal int HitCount => _hitCount;
+
+            private int _missCount = 0;
+            internal int MissCount => _missCount;
+
+            private int _cacheBypassCount = 0;
+            internal int CacheBypassCount => _cacheBypassCount;
+
+            private int _expiredCacheCount = 0;
+            internal int ExpiredCacheCount => _expiredCacheCount;
+
+            internal void IncrementCacheHitCount()
+            {
+                Interlocked.Increment(ref _hitCount);
+            }
+
+            internal void IncrementCacheMissCount(bool cacheByPass, bool isExpired)
+            {
+                Interlocked.Increment(ref _missCount);
+
+                if (cacheByPass)
+                {
+                    Interlocked.Increment(ref _cacheBypassCount);
+                }
+
+                if (isExpired)
+                {
+                    Interlocked.Increment(ref _expiredCacheCount);
+                }
+            }
+
+            internal void AddProperties(TelemetryEvent telemetryEvent)
+            {
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.HitCount] = HitCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.MissCount] = MissCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.CacheBypassCount] = CacheBypassCount;
+                telemetryEvent[HttpCache + NuGetHttpCacheUsageCollector.ExpiredCacheCount] = ExpiredCacheCount;
+            }
```

Comment:
I agree with @zivkan.

I'm also not sure why this needs a separate event.
http cache hits/misses are on a per project level, so if they were being sent,t hey should be sent through protocol diagnostics.

I do think this information can be valuable, and we might learn something from it, but I don't think it addresses the issue it claims to address.

Reply:
> I'm also not sure why this needs a separate event.
> http cache hits/misses are on a per project level, so if they were being sent,t hey should be sent through protocol diagnostics.

The PR is (now) raising .NET events in the same way that other ProtocolDiagnostics events are (and it already has multiple events). So, FWIW, I'm not 100% sure what you're referring to.

If you mean why is a new .NET event added to the `ProtocolDiagnostics` class, it's consistent with the existing ProtocolDiagnostics architecture. When there's a cache hit `RaiseEvent(ProtocolDiagnosticHttpEvent)` is never called, because it never calls into HttpRetryHandler. `RaiseEvent(ProtocolDiagnosticResourceEvent)` should be called on all SourceRepository resource public APIs, but in order to add cache hit/miss info to those events, HttpSource would have to pass the information all the way up the call stack (and previous comments on this PR already, rightfully, discouraged changs to public APIs.  ProtocolDiagnostics already has 3 different .NET events for this reason - passing data up and down the callstack is difficult and horrible for public APIs, so different events are used.

Reply:
I get why there's a new .NET event.

Why shouldn't this be sent with PackageSourceTelemetry?

Reply:
> 

I'll check if I can add into SourceTelemetry without breaking public apis.

Reply:
Personally, I'm not worried about the new public APIs, that seems ok no? 

My comment is mostly whether they should be included in an event about the source usage during that operation.

Reply:
Ah, I get it now. This PR is adding a property to the `vs/nuget/instanceclose` event, but you think it should be added to the `vs/nuget/packagesourcediagnostic` event: https://github.com/NuGet/NuGet.Client/blob/0e3a595fff9f6c0be61132b2771f8401e112b47a/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs#L220-L255

Yes, that makes sense. We're already collecting a few HTTP related properties there, so it makes sense for us to add it there. Plus, it will allow us to see cache hit/miss differences between PM UI and restore.

However, this "conversation"/thread started because the PR template says that this will help us understand whether `If-Modified-Since` should be implemented, but this PR won't help with that.

Reply:
@nkolev92 @zivkan 
I did move `http cache logic` to HTTP related properties in `PackageSourceTelemetry`. 
Currently this telemetry is emitted from only 2 places, 1st from [a solution restore job](https://github.com/NuGet/NuGet.Client/blob/f5d7dc371d8a54a9fe0869698e5c4f29dcb7d981/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs#L115) and 2nd from [pmui search](https://github.com/NuGet/NuGet.Client/blob/f5d7dc371d8a54a9fe0869698e5c4f29dcb7d981/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageFeed.cs#L115), both of doesn't happen frequently, so works fine as they happen sequentially.
But it doesn't cover PMUI meta data requests in Install/Update tabs (which consist of hundreds of Http calls), so undercounting one of heavy usage area, fixing this would be quite a lot of refactoring in many places, ROI would be not great, also considering some PMUI design change happening now it's not good timing. I tried to add `PackageSourceTelemetry` in [MultiSourcePackageMetadataProvider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs#L256-L267) (which is closest possible place to UI), as [you can see](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs#L256-L267) each `packageId` meta data requests are handled separately and in it making concurrent http calls to all sources which makes it impossible to use it with `PackageSourceTelemetry`.
Here is example: Let's say Project has following dependencies `PackageA`, `PackageB`, `PackageC` etc..
`PackageA` creates new `PackageSourceTelemetry` then register own [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/dev-eryondon-addcache-hit-miss/src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/PackageSourceTelemetry.cs#L67).
Same time `PackageB`, `PackageC` the register to same event same time.  Now everytime [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/563d1f2e9f110081b1394e41f9a854e39113f1f4/src/NuGet.Core/NuGet.Protocol/Events/ProtocolDiagnostics.cs#L41) is triggered it send data to all 3 instances of `PackageSourceTelemetry` therefore inflate numbers to all instances listening to it.

So, proposal is "revert back to previous implementation (collect all then send data when close VS, see screenshot in description area)", but I'll group numbers by source and add hash change detection. Does it sound workable solution for you?

Reply:
http communication is something that NuGet always does globally. 

I am not sure I follow your concern about double counting. Why is that particular to package source diagnostics and not to any other implementation? 

@zivkan also brought up a very interesting point: 

> However, this "conversation"/thread started because the PR template says that this will help us understand whether If-Modified-Since should be implemented, but this PR won't help with that.

which I agree with. This just captures cache hit/miss but does not really tell us whether something actually changed. 

 > ROI would be not great, also considering some PMUI design change happening now it's not good timing
 
 I think the same logic applies to every attempt to track the PM UI behavior, which what you're trying to do regardless of the implementation. I get that it's slightly different because it'll report it at the end of the process, but I don't think it carries much value when implemented in such a way.

Reply:
Putting aside the issue that this PR will not help the scenarios that the PR description claims

> fixing this would be quite a lot of refactoring in many places, ROI would be not great

On the other hand, HTTP timeouts are NuGet's number 1 unhandled exception in VS, so improving this would benefit not only the telemetry you want to add, but also help us understand what customer experiences are being harmed by HTTP timeouts.

> Now everytime [HttpCacheEvent](https://github.com/NuGet/NuGet.Client/blob/563d1f2e9f110081b1394e41f9a854e39113f1f4/src/NuGet.Core/NuGet.Protocol/Events/ProtocolDiagnostics.cs#L41) is triggered it send data to all 3 instances of PackageSourceTelemetry therefore inflate numbers to all instances listening to it.

This suggests that you should add `string PackageSource { get; set;}` to to `HttpCacheEvent`, so when the event happens, you know which source it happened in, and you can filter in the event listener. Isn't that what all the other PackageSourceDiagnostics events do?

Reply:
> ~This suggests that you should add `string PackageSource { get; set;}` to to `HttpCacheEvent`, so when the event happens, you know which source it happened in, and you can filter in the event listener. Isn't that what all the other PackageSourceDiagnostics events do?~
> 
> No, the code you added already filters to a single source: https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-cd3abb581fb4e39b8057c9bb2da526ee0a16bb3371a049b375e4552ccfb2984aR199-R202

Main problem is how PMUI acquire meta data here, PMUI get meta data request made 1 by 1 for per package through http starting from PackageItemViewModel level. So now each `packageId + source` is new event, not only they interfere each other, also increase number of `PackageSourceTelemetry` many multifold and spam telemetry service.

Reply:
@andy @nkolev92 
Yesterday, I accidentally discovered meta data path such as `https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json` are not well tested/supported by nuget.org, by playing with `If-Modified-Since` support for nuget.org in order to detect meta data content change. 

The CDN edge node for [my ISP sometimes doesn't honor](https://github.com/NuGet/NuGetGallery/issues/9175) `If-Modified-Since`, `Etag` header (@joelverhagen can repro too). So, adding content change detection is not reliable at this point with `If-Modified-Since`, `Etag` headers. Fixing that is not high priority for server team right now, if there's any data to prioritize it then they can do it.

At this moment could scope effort on `http-cache bypass/expire/hit/miss` count telemetry for PMUI without content change detection? 

Actually, we can do meta data content change frequency analysis directly on nuget.org, ofc there could be some data correlation issue depending on analysis conclusion.

What we can achieve with this `http-cache intentional bypass/expire/hit/miss` count is:
1. If cache `miss` happening without most of the time without `intentional bypass` then not much we can do, no follow up action.
2. If ton of cache `miss` happening with `intentional bypass` then we should consider making use of existing cache on disc with combination of `If-Modified-Since` or `Etag`. For this we could ask server team to fix it.
3. If a lot of `hit` and `expire` is happening then we should consider [increase cache file life](https://github.com/NuGet/Home/issues/3390) (currently just 30 min) or with combination of `If-Modified-Since` or `Etag` just update the modified datetime of existing cache on disc.  Currently it deletes the existing cache file after 30 min considering it as expired then download exact same file to replace it. For this we could ask server team to fix it.
4. If a lot of hit without expire then probably not much room for improvement. It means most calls are made to different endpoints, caching doesn't work well here.

Reply:
> Yesterday, I accidentally discovered meta data path such as https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json are not well tested/supported by nuget.org, by playing with If-Modified-Since support for nuget.org in order to detect meta data content change.

I'm not sure I understand what this means. Registration pages are very much supported and likely tested. 
I guess you mean to say the `If-Modified-Since` header isn't? 

> At this moment could scope effort on http-cache bypass/expire/hit/miss count telemetry for PMUI without content change detection?

That sound good, but it's only 1 of the questions we had. 
The `scope` of the data collected matters too. A few of us voiced feedback about collecting session telemetry only and that it's not helpful. 

> What we can achieve with this http-cache intentional bypass/expire/hit/miss count is:

Thanks for the summary, I think that's reasonable, something that probably should have probably been a telemetry design as @dominoFire suggested.

> If a lot of hit and expire is happening then we should consider https://github.com/NuGet/Home/issues/3390 (currently just 30 min) or with combination of If-Modified-Since or Etag just update the modified datetime of existing cache on disc. Currently it deletes the existing cache file after 30 min considering it as expired then download exact same file to replace it. For this we could ask server team to fix it.

When we make assumptions about what a server can do, we *have* to account for the fact that nuget.org is not the only source. Our solutions need can't rely on nuget.org behavior only.

Reply:
> > Yesterday, I accidentally discovered meta data path such as https://api.nuget.org/v3/registration5-gz-semver2/stackexchange.redis/index.json are not well tested/supported by nuget.org, by playing with If-Modified-Since support for nuget.org in order to detect meta data content change.
> 
> I'm not sure I understand what this means. Registration pages are very much supported and likely tested. I guess you mean to say the `If-Modified-Since` header isn't?
> 
Yes, I mean `If-Modified-Since` header is not supported well, details are [here](https://github.com/NuGet/NuGetGallery/issues/9175).

> > At this moment could scope effort on http-cache bypass/expire/hit/miss count telemetry for PMUI without content change detection?
> 
> That sound good, but it's only 1 of the questions we had. The `scope` of the data collected matters too. A few of us voiced feedback about collecting session telemetry only and that it's not helpful.
> 

Sure, I'll reevaluate if need to collect new telemetry, technically I already implemented `If-Modified-Since` [here](https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-fe623b43eb533767f04846c3aea5266985147b70a31c2386d7bd8d9275991f63R137-R141). So, this moment just go ahead with implementation without adding new telemetry could be cheaper, but there is some edge case concern, so I'm thinking about writing actual spec for implementing `If-Modified-Since`. Actually, when Joel tested `If-Modified-Since` worked as expected for 20/20. It looks like some CDN edge nodes are misbehaving.

> When we make assumptions about what a server can do, we _have_ to account for the fact that nuget.org is not the only source. Our solutions need can't rely on nuget.org behavior only.

I know many smaller org want to see new packages immediately without caching, so I was not planning to collect data for them. My goal was NuGet.org for initial step extend to Azure and others gradually if it works for NuGet.org. 

I'll step back and reevaluate everything 1 more time.

Reply:
> Sure, I'll reevaluate if need to collect new telemetry, technically I already implemented If-Modified-Since [here](https://github.com/NuGet/NuGet.Client/pull/4668/files#diff-fe623b43eb533767f04846c3aea5266985147b70a31c2386d7bd8d9275991f63R137-R141). So, this moment just go ahead with implementation without adding new telemetry could be cheaper, but there is some edge case concern, so I'm thinking about writing actual spec for implementing If-Modified-Since. Actually, when Joel tested If-Modified-Since worked as expected for 20/20. It looks like some CDN edge nodes are misbehaving.

As long as we're not getting false negative we can consider it. 

I'd be ok with implementing the header. I'm likely to consider adding it when working on the vulnerabities in restore.

> I know many smaller org want to see new packages immediately without caching, so I was not planning to collect data for them. My goal was NuGet.org for initial step extend to Azure and others gradually if it works for NuGet.org.

That wasn't really my point. We can make protocol improvements and we can tweak nuget.org, but when we look at our client perf, we can't make any of those assumptions unless we have a special case such as vulnerabilities, because they are *only* reported in nuget.org and we can define support for IF-Modified-Since on that potential new resource as a requiement.